File: app/src/test/java/com/beautifulquran/ui/rootviewer/MorphologyLabelsTest.kt
package com.beautifulquran.ui.rootviewer

import org.junit.Assert.assertEquals
import org.junit.Test

class MorphologyLabelsTest {

    @Test
    fun `pos labels cover common tags`() {
        assertEquals("Noun", MorphologyLabels.posLabel("N"))
        assertEquals("Verb", MorphologyLabels.posLabel("V"))
        assertEquals("Proper noun", MorphologyLabels.posLabel("PN"))
        assertEquals("XYZ", MorphologyLabels.posLabel("XYZ"))
    }

    @Test
    fun `feature summary joins known tags`() {
        assertEquals(
            "perfect · masculine singular · genitive",
            MorphologyLabels.featureSummary("PERF|MS|GEN"),
        )
        assertEquals("form X", MorphologyLabels.featureSummary("(X)"))
        assertEquals("", MorphologyLabels.featureSummary(""))
    }

    @Test
    fun `feature summary decodes explicit corpus mood and definiteness tags`() {
        assertEquals(
            "imperfect · indefinite · jussive mood · 3rd person masculine dual",
            MorphologyLabels.featureSummary("IMPF|INDEF|MOOD:JUS|3MD"),
        )
    }

    @Test
    fun `spaced root inserts letter gaps`() {
        assertEquals("ك ت ب", MorphologyLabels.spacedRoot("كتب"))
        assertEquals("س م و", MorphologyLabels.spacedRoot("سمو"))
    }
}

File: web/src/ui/root/__tests__/morphologyLabels.test.ts
import { describe, expect, it } from 'vitest'
import { featureSummary, posLabel, spacedRoot } from '../morphologyLabels'

describe('posLabel', () => {
  it('covers common tags', () => {
    expect(posLabel('N')).toBe('Noun')
    expect(posLabel('V')).toBe('Verb')
    expect(posLabel('PN')).toBe('Proper noun')
    expect(posLabel('XYZ')).toBe('XYZ')
  })
})

describe('featureSummary', () => {
  it('joins known tags', () => {
    expect(featureSummary('PERF|MS|GEN')).toBe('perfect · masculine singular · genitive')
    expect(featureSummary('(X)')).toBe('form X')
    expect(featureSummary('')).toBe('')
  })

  it('decodes explicit corpus mood and definiteness tags', () => {
    expect(featureSummary('IMPF|INDEF|MOOD:JUS|3MD')).toBe(
      'imperfect · indefinite · jussive mood · 3rd person masculine dual',
    )
  })

  it('drops unrecognized tags and de-duplicates', () => {
    expect(featureSummary('PERF|NOPE|PERF')).toBe('perfect')
  })
})

describe('spacedRoot', () => {
  it('inserts letter gaps', () => {
    expect(spacedRoot('كتب')).toBe('ك ت ب')
    expect(spacedRoot('سمو')).toBe('س م و')
  })
})
